// @magnolia/cli-create-light-module-plugin/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import fs from 'fs-extra';
import inquirer from 'inquirer';
import { availableDirectories } from './mapping.js';
import { i18nInstance, logger } from '../create-light-module-plugin.js';
export function filterDirectories(srcPath, destPath, directories) {
    return __awaiter(this, void 0, void 0, function* () {
        let allDirectories = yield getDirectories(srcPath);
        const notFoundDirs = [];
        let selectedDirs = [];
        if (Array.isArray(directories)) {
            directories.forEach((dir) => {
                let findDir;
                if (dir === dir.toUpperCase()) {
                    findDir = availableDirectories[dir];
                }
                else {
                    findDir = allDirectories.find((d) => d === dir);
                }
                if (findDir) {
                    selectedDirs.push(findDir);
                }
                else {
                    notFoundDirs.push(dir);
                }
            });
            selectedDirs = filterMissingDirs(selectedDirs, destPath);
        }
        if (directories === true || notFoundDirs.length > 0) {
            if (notFoundDirs.length > 0) {
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-not-found-directories', {
                    directories: notFoundDirs.join(', '),
                }));
            }
            if (fs.existsSync(destPath)) {
                allDirectories = filterMissingDirs(allDirectories, destPath);
            }
            if (allDirectories.length > 0) {
                const answer = yield inquirer.prompt([
                    {
                        type: 'checkbox',
                        name: 'directories',
                        message: 'Please select the directories that you want to be created',
                        choices: allDirectories,
                        default: selectedDirs,
                    },
                ]);
                selectedDirs = answer.directories;
            }
        }
        return { selectedDirs, notFoundDirs };
    });
}
export function filterMissingDirs(src, dest) {
    if (typeof src === 'string') {
        src = fs.readdirSync(src);
    }
    if (!fs.existsSync(dest)) {
        return src;
    }
    const destDirs = fs.readdirSync(dest);
    return src.filter((dir) => !destDirs.includes(dir));
}
const getDirectories = (source) => __awaiter(void 0, void 0, void 0, function* () {
    return (yield fs.readdir(source, { withFileTypes: true }))
        .filter((dirent) => dirent.isDirectory())
        .map((dirent) => dirent.name);
});
