// @magnolia/cli-create-page-plugin/dist/cli-create-page.js
import path from 'path';
import { fileURLToPath } from 'url';
import TemplateHelperPlugin from '@magnolia/cli-template-helper';
import { createRequire } from 'module';
import { Option } from 'commander';
import { initI18n } from '@magnolia/cli-helper/i18n';
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
export default class CreatePagePlugin extends TemplateHelperPlugin {
    constructor(args) {
        super(args, 'pages');
        this.name = 'create-page';
        this.usage = '[name] [options]' + '\nParameter: [name] The name of the page.';
        this.version = pkg.version;
        const i18nInstance = initI18n(this.name, 'translation', path.join(__dirname, 'lib/locales'));
        this.description = i18nInstance.t('description');
        this.options = [
            new Option('-lm, --light-module [name]', i18nInstance.t('option-lightModule-description')),
            new Option('-p, --prototype [name]', i18nInstance.t('option-prototype-description')),
        ];
    }
}
