// @magnolia/cli-generate-plugin/dist/generate-plugin.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PluginTemplate } from '@magnolia/cli-plugin-template';
import { Option } from 'commander';
import { createRequire } from 'module';
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
import { resolveOptionOrPrompt, isValidClassName, isValidPluginName, initI18n, } from './lib/helper.js';
import { generateNewPlugin } from './lib/generate.js';
export let logger;
export let i18nInstance = {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    t(key, options) {
        return key;
    },
};
export default class GeneratePlugin extends PluginTemplate {
    constructor() {
        super();
        this.name = 'generate-plugin';
        this.version = pkg.version;
        this.usage = '[options]';
        i18nInstance = initI18n(this.name);
        this.description = i18nInstance.t('description');
        this.options = [
            new Option('-n, --name <name>', i18nInstance.t('option-name-description')),
            new Option('-c, --class <name>', i18nInstance.t('option-class-description')),
        ];
    }
    start(options) {
        return __awaiter(this, void 0, void 0, function* () {
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-will-be-in-cwd'));
            const templateVariables = {
                pluginName: yield resolveOptionOrPrompt(i18nInstance.t('inquirer-prompt-name'), 'cli-test-plugin', isValidPluginName, options.name),
                className: yield resolveOptionOrPrompt(i18nInstance.t('inquirer-prompt-class'), 'TestPlugin', isValidClassName, options.class),
            };
            yield generateNewPlugin(templateVariables);
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-develop'));
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-modify', {
                pluginName: templateVariables.pluginName,
            }));
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-more-information'));
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    init(winstonLogger) {
        return __awaiter(this, void 0, void 0, function* () {
            logger = winstonLogger;
        });
    }
}
