// @magnolia/cli/dist/lib/analytics.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import axios from 'axios';
import { v4 as uuidv4, validate, version } from 'uuid';
import fs from 'fs-extra';
import * as os from 'os';
import { execSync } from 'node:child_process';
import { logger } from '../Cli.js';
import { BEAUTIFY_OPTIONS } from '@magnolia/cli-helper/general-utils';
import beautifyModule from 'js-beautify';
const { js_beautify } = beautifyModule;
const METRICS_URL = 'https://usage-metrics.magnolia-cms.com/';
function send(cliVersion, config, plugin) {
    return __awaiter(this, void 0, void 0, function* () {
        const analytics = config === null || config === void 0 ? void 0 : config.analytics;
        if (analytics === undefined ||
            !(analytics === null || analytics === void 0 ? void 0 : analytics.enabled) ||
            !(validate(analytics === null || analytics === void 0 ? void 0 : analytics.uuid) && version(analytics === null || analytics === void 0 ? void 0 : analytics.uuid)))
            return;
        try {
            const { osName, osVersion } = getCorrectOSInfo();
            const data = {
                osArch: os.arch(),
                osVersion: osVersion,
                osName: osName,
                sites: [], // Note: Required by the metric endpoint (not tracked)
                locales: [], // Note: Required by the metric endpoint (not tracked)
                modules: [], // Note: Required by the metric endpoint (not tracked)
                develop: true, // Note: Required by the metric endpoint (not tracked)
                uuid: analytics.uuid,
                nodeVersion: process.version,
                cliVersion: cliVersion,
                command: { name: plugin.name, version: plugin.version },
            };
            yield axios.post(METRICS_URL, data, {
                headers: {
                    'Content-Type': 'application/json',
                },
            });
        }
        catch (_a) {
            logger === null || logger === void 0 ? void 0 : logger.warn("Couldn't send analytics.");
        }
    });
}
function init(configPath) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!fs.existsSync(configPath))
            return;
        try {
            const customConfigContent = fs.readFileSync(configPath).toString();
            // eslint-disable-next-line no-useless-escape
            const analyticsPattern = /(analytics\s*:\s*)(\{)([^\}]*?)(\})/;
            if (!customConfigContent.match(analyticsPattern)) {
                const exportDefaultPattern = /(export default \{)([\s\S]*?)(\}(?![^}]*});?)/;
                const analyticsStr = `analytics: {\nenabled:true, // Set to false to turn off analytics\n uuid: "${uuidv4()}"\n}`;
                const updatedConfigContent = customConfigContent.replace(exportDefaultPattern, (_, p1, p2, p3) => {
                    const sep = p2.trim() !== '' && analyticsStr.trim() !== ''
                        ? ','
                        : '';
                    return `${p1}${analyticsStr}${sep}${p2}${p3}`;
                });
                fs.writeFileSync(configPath, js_beautify(updatedConfigContent.trim(), BEAUTIFY_OPTIONS));
                logger === null || logger === void 0 ? void 0 : logger.info('We collect anonymised data to improve your experience.\n' +
                    'Data include the names and versions of used commands, Magnolia CLI version, Node.js version and OS information.\n' +
                    'We use this data to prioritize further development of the tool.\n' +
                    'All collected data are collected anonymously and not tied to any individual.\n' +
                    'No personal information is collected.');
                logger === null || logger === void 0 ? void 0 : logger.info("If you prefer to opt out, set 'analytics.enabled' to 'false' in mgnl.config.js file.\n");
            }
        }
        catch (_a) {
            logger === null || logger === void 0 ? void 0 : logger.warn('Cannot set analytics object.');
        }
    });
}
function getCorrectOSInfo() {
    let osName = os.platform().toString();
    let osVersion = os.release();
    switch (os.platform()) {
        case 'darwin':
            osName = 'Mac OS X';
            osVersion = execSync('sw_vers -productVersion').toString().trim();
            break;
        case 'win32':
            osName = 'Windows';
            break;
        case 'linux':
            osName = 'Linux';
            break;
    }
    return { osName, osVersion };
}
export const analytics = { send, init };
