// @magnolia/cli/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import ora from 'ora';
import { execa } from 'execa';
import fs from 'fs-extra';
import { initI18n } from '@magnolia/cli-helper/i18n';
import { fileURLToPath } from 'url';
import path from 'path';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const i18nInstance = initI18n('helper', 'helper', path.join(__dirname, 'locales'));
export const hasCliInstalled = (packageJson) => {
    var _a, _b;
    return (((_a = packageJson === null || packageJson === void 0 ? void 0 : packageJson.dependencies) === null || _a === void 0 ? void 0 : _a['@magnolia/cli']) !== undefined ||
        ((_b = packageJson === null || packageJson === void 0 ? void 0 : packageJson.devDependencies) === null || _b === void 0 ? void 0 : _b['@magnolia/cli']) !== undefined);
};
export const installCLI = () => __awaiter(void 0, void 0, void 0, function* () {
    const packageManager = determinePackageManager();
    const installFlag = packageManager === 'npm' ? '--save-dev' : '--dev';
    return yield installPackage(packageManager, '@magnolia/cli@latest', '@magnolia/cli', [installFlag]);
});
export const installPackage = (packageManager, packageReference, packageName, extraFlags) => __awaiter(void 0, void 0, void 0, function* () {
    if (!packageName) {
        packageName = packageReference;
    }
    const spinner = ora().start(i18nInstance.t('ora-start-installing-package', {
        packageName: packageName,
    }));
    try {
        const installCommand = packageManager === 'npm' ? 'install' : 'add';
        const args = [installCommand, packageReference];
        if (extraFlags && extraFlags.length > 0) {
            args.push(...extraFlags);
        }
        yield execa(packageManager, args, {
            buffer: true,
            cwd: process.cwd(),
        });
        spinner.succeed(i18nInstance.t('ora-succeed-package-installed', {
            packageName: packageName,
        }));
        return true;
    }
    catch (error) {
        spinner.fail(i18nInstance.t('ora-fail-installing-package', {
            packageName: packageName,
            errorMsg: error.message,
        }));
        return false;
    }
});
export const determinePackageManager = (baseDirectory = process.cwd()) => {
    const yarnLockPath = path.join(baseDirectory, 'yarn.lock');
    const packageLockPath = path.join(baseDirectory, 'package-lock.json');
    if (fs.existsSync(yarnLockPath)) {
        return 'yarn';
    }
    else if (fs.existsSync(packageLockPath)) {
        return 'npm';
    }
    return 'npm';
};
