// @magnolia/cli/dist/lib/tab-completion.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PluginTemplate } from '@magnolia/cli-plugin-template';
import { Option } from 'commander';
import path from 'path';
import fs from 'fs-extra';
import os from 'node:os';
import { initI18n } from '@magnolia/cli-helper/i18n';
import { fileURLToPath } from 'url';
import { getMgnlConfig, getMgnlConfigPath, modifyConfig, } from '@magnolia/cli-helper';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
import inquirer from 'inquirer';
import { ExitPromptError } from '@inquirer/core';
import { execa } from 'execa';
import Handlebars from 'handlebars';
import { execSync } from 'node:child_process';
const i18nNamespace = 'tab-completion';
class TabCompletionError extends Error {
    constructor(message) {
        super(message);
        this.message = message;
    }
}
export default class TabCompletion extends PluginTemplate {
    constructor(args) {
        super();
        this.name = 'tab-completion';
        this.usage = '[options]';
        this.paths = {
            MGNL_V5_DIR_PATH: path.join(os.homedir(), '.mgnl_v5'),
            TAB_COMP_SCRIPT_PATH: path.join(os.homedir(), '.mgnl_v5', os.platform() === 'win32' ? 'tab_completion.ps1' : 'tab_completion'),
            WINDOWS_USER_PROFILE_PATHS: [
                path.join(os.homedir(), 'Documents', 'WindowsPowerShell', 'profile.ps1'),
            ],
            USER_SHELL_CONFIG_PATHS: [
                path.join(os.homedir(), '.bashrc'),
                path.join(os.homedir(), '.bash_profile'),
                path.join(os.homedir(), '.profile'),
                path.join(os.homedir(), '.zshrc'),
            ],
        };
        this.markers = {
            SHELL_AND_PROFILE_START: '# mgnl-v5-tabcompletion-loader-start',
            SHELL_AND_PROFILE_END: '# mgnl-v5-tabcompletion-loader-end',
            TAB_COMPLETION_START: '# mgnl-v5-tabcompletion-start',
            TAB_COMPLETION_END: '# mgnl-v5-tabcompletion-end',
        };
        this.systemScriptsPaths = os.platform() === 'win32'
            ? this.paths.WINDOWS_USER_PROFILE_PATHS
            : this.paths.USER_SHELL_CONFIG_PATHS;
        this.i18nInstance = initI18n(this.name, i18nNamespace, path.join(__dirname, 'locales'));
        this.description = this.i18nInstance.t('description');
        this.version = args.version;
        this.options = [
            new Option('-e, --enable', this.i18nInstance.t('option-enable-description')).conflicts(['disable', 'uninstall', 'reinstall', 'reset-cache']),
            new Option('-d, --disable', this.i18nInstance.t('option-disable-description'))
                .conflicts('enable')
                .default(false),
            new Option('-u, --uninstall', this.i18nInstance.t('option-uninstall-description'))
                .conflicts(['reinstall', 'reset-cache', 'enable'])
                .default(false),
            new Option('-r, --reinstall', this.i18nInstance.t('option-reinstall-description'))
                .conflicts(['reset-cache', 'uninstall', 'enable'])
                .default(false),
            new Option('-rc, --reset-cache', this.i18nInstance.t('option-reset-cache-description'))
                .conflicts(['reinstall', 'uninstall', 'enable'])
                .default(false)
                .hideHelp(),
        ];
    }
    start(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            try {
                if (options.resetCache || options.reinstall || options.uninstall) {
                    yield this.resetCache();
                    if (options.resetCache) {
                        return;
                    }
                }
                if (options.uninstall) {
                    yield this.uninstall();
                    yield this.toggle(false);
                    return;
                }
                yield this.install(options.reinstall);
                yield this.toggle(!options.disable);
            }
            catch (e) {
                if (e instanceof ExitPromptError)
                    return;
                if (e instanceof TabCompletionError) {
                    (_a = this.logger) === null || _a === void 0 ? void 0 : _a.error(e.message);
                }
                else if (e instanceof Error) {
                    (_b = this.logger) === null || _b === void 0 ? void 0 : _b.error(this.i18nInstance.t('error-non-mgnl-error', {
                        errorMsg: e.message,
                    }));
                }
                else {
                    throw e;
                }
            }
        });
    }
    stop() {
        return Promise.resolve(undefined);
    }
    init(logger) {
        this.logger = logger;
    }
    toggle() {
        return __awaiter(this, arguments, void 0, function* (enabled = true) {
            var _a, _b, _c;
            const localMgnlConfig = yield getMgnlConfig(this.logger);
            const localMgnlConfigPath = getMgnlConfigPath();
            if (!fs.existsSync(localMgnlConfigPath)) {
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-toggle-use-in-project'));
                return;
            }
            if (localMgnlConfig.tabCompletion !== enabled) {
                yield modifyConfig({ sharedProps: { tabCompletion: enabled } }, '', this.logger);
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(this.i18nInstance.t('info-toggle-update', {
                    enabled: enabled ? 'enabled' : 'disabled',
                }));
            }
            else {
                (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(this.i18nInstance.t('info-toggle', {
                    enabled: enabled ? 'enabled' : 'disabled',
                }));
            }
        });
    }
    install() {
        return __awaiter(this, arguments, void 0, function* (reinstall = false) {
            const v4Configured = yield this.isV4Configured();
            yield this.checkDependencies();
            fs.ensureDirSync(this.paths.MGNL_V5_DIR_PATH);
            yield this.ensureTabCompletionScript(this.paths.TAB_COMP_SCRIPT_PATH, v4Configured, reinstall);
            this.configureTabCompletionLoader();
        });
    }
    isV4Configured() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c;
            const v4scriptExists = fs.existsSync(path.join(os.homedir(), '.mgnl', 'mgnl'));
            const v4Set = [];
            this.systemScriptsPaths.forEach((path) => {
                if (fs.existsSync(path)) {
                    const content = fs.readFileSync(path).toString();
                    v4Set.push(content.includes('mgnl-tabcompletion-start'));
                }
            });
            if (v4Set.some(Boolean) || v4scriptExists) {
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-v4-configured'));
                const { confirm } = yield inquirer.prompt({
                    type: 'confirm',
                    name: 'confirm',
                    message: this.i18nInstance.t('inquirer-confirm-v4-removal'),
                    default: true,
                });
                if (confirm) {
                    yield execa('npx', ['@magnolia/cli@4', 'tab-completion', 'uninstall'], { cwd: process.cwd() });
                    const v4scriptExists2 = fs.existsSync(path.join(os.homedir(), '.mgnl', 'mgnl'));
                    const v4Set2 = [];
                    this.systemScriptsPaths.forEach((path) => {
                        if (fs.existsSync(path)) {
                            const content = fs.readFileSync(path).toString();
                            v4Set2.push(content.includes('mgnl-tabcompletion-start'));
                        }
                    });
                    if (!v4Set2.some(Boolean) && !v4scriptExists2) {
                        return false;
                    }
                    (_b = this.logger) === null || _b === void 0 ? void 0 : _b.error(this.i18nInstance.t('error-v4-removal'));
                }
                else {
                    (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(this.i18nInstance.t('info-v4-not-removed'));
                }
                return true;
            }
            return false;
        });
    }
    checkDependencies() {
        return __awaiter(this, void 0, void 0, function* () {
            const isWin = os.platform() === 'win32';
            const dependencies = isWin
                ? ['node', 'npm']
                : ['node', 'jq', 'npm', 'grep', 'sed', 'awk', 'stat'];
            const commandExists = (cmd) => {
                try {
                    if (isWin) {
                        execSync(`where ${cmd}`, { stdio: 'ignore' });
                    }
                    else {
                        execSync(`command -v ${cmd}`, { stdio: 'ignore' });
                    }
                    return true;
                }
                catch (_a) {
                    return false;
                }
            };
            const missing = dependencies.filter((cmd) => !commandExists(cmd));
            if (missing.length > 0) {
                throw new TabCompletionError(this.i18nInstance.t('error-missing-deps', {
                    deps: missing.join(', '),
                }));
            }
        });
    }
    ensureTabCompletionScript(scriptPath_1) {
        return __awaiter(this, arguments, void 0, function* (scriptPath, v4Configured = false, reinstall = false) {
            var _a, _b, _c;
            const templateFile = path.join(__dirname, '..', 'resources', 'tab-completion', `script.${os.platform() === 'win32' ? 'ps1' : 'sh'}.hbs`);
            const templateSource = fs.readFileSync(templateFile, 'utf8');
            const template = Handlebars.compile(templateSource);
            const buildScriptContent = (context) => {
                return (this.markers.TAB_COMPLETION_START +
                    '\n' +
                    template(context) +
                    '\n' +
                    this.markers.TAB_COMPLETION_END);
            };
            const validVariants = [
                buildScriptContent({ local: true, global: false }).trim(),
                buildScriptContent({ local: false, global: true }).trim(),
                buildScriptContent({ local: true, global: true }).trim(),
            ];
            if (!reinstall) {
                if (fs.existsSync(scriptPath)) {
                    const currentContent = fs
                        .readFileSync(scriptPath, 'utf8')
                        .trim();
                    if (validVariants.includes(currentContent)) {
                        return;
                    }
                    const { confirm } = yield inquirer.prompt({
                        type: 'confirm',
                        name: 'confirm',
                        message: this.i18nInstance.t('inquirer-confirm-script-recreation', { path: scriptPath }),
                        default: true,
                    });
                    if (!confirm) {
                        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.warn(this.i18nInstance.t('info-recreate-script-future'));
                        return;
                    }
                }
            }
            const { installationType } = yield inquirer.prompt([
                {
                    type: 'list',
                    name: 'installationType',
                    message: this.i18nInstance.t('inquirer-choose-install-type'),
                    choices: [
                        {
                            name: this.i18nInstance.t('inquirer-choose-install-type-local'),
                            value: 'local',
                        },
                        ...(!v4Configured
                            ? [
                                {
                                    name: this.i18nInstance.t('inquirer-choose-install-type-global'),
                                    value: 'global',
                                },
                                {
                                    name: this.i18nInstance.t('inquirer-choose-install-type-both'),
                                    value: 'both',
                                },
                            ]
                            : []),
                    ],
                },
            ]);
            const context = {
                local: installationType === 'local' || installationType === 'both',
                global: installationType === 'global' || installationType === 'both',
            };
            const scriptContent = buildScriptContent(context);
            if (fs.existsSync(scriptPath)) {
                fs.writeFileSync(scriptPath, scriptContent, 'utf8');
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(this.i18nInstance.t('info-script-recreated', {
                    path: scriptPath,
                }));
            }
            else {
                fs.writeFileSync(scriptPath, scriptContent, 'utf8');
                (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(this.i18nInstance.t('info-script-created', {
                    path: scriptPath,
                }));
            }
        });
    }
    getTabCompletionLoaderScript(content, filePath) {
        let configScript = this.markers.SHELL_AND_PROFILE_START + '\n';
        if (os.platform() === 'win32') {
            configScript += '. "$HOME\\.mgnl_v5\\tab_completion.ps1"\n';
        }
        else {
            if (filePath.endsWith('.zshrc')) {
                configScript += 'autoload bashcompinit\nbashcompinit\n';
            }
            configScript += 'source ~/.mgnl_v5/tab_completion\n';
        }
        configScript += this.markers.SHELL_AND_PROFILE_END + '\n';
        if (!content.endsWith('\n') && content !== '') {
            configScript = `\n${configScript}`;
        }
        return configScript;
    }
    configureTabCompletionLoader() {
        var _a, _b;
        const success = [];
        let wasInstalledBefore = false;
        let selectedConfigFile = null;
        if (os.platform() === 'win32') {
            selectedConfigFile = this.paths.WINDOWS_USER_PROFILE_PATHS[0];
            if (!fs.existsSync(selectedConfigFile)) {
                fs.writeFileSync(selectedConfigFile, '', { encoding: 'utf-8' });
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-created-file', {
                    path: selectedConfigFile,
                }));
            }
        }
        else {
            selectedConfigFile =
                this.paths.USER_SHELL_CONFIG_PATHS.find(fs.existsSync) || null;
            if (!selectedConfigFile) {
                const userShell = process.env.SHELL || '';
                if (userShell.includes('zsh')) {
                    selectedConfigFile = path.join(os.homedir(), '.zshrc');
                }
                else {
                    selectedConfigFile = path.join(os.homedir(), '.bashrc');
                }
                fs.writeFileSync(selectedConfigFile, '', { encoding: 'utf-8' });
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(this.i18nInstance.t('info-created-file', {
                    path: selectedConfigFile,
                }));
            }
        }
        if (selectedConfigFile && this.canWriteToConfig(selectedConfigFile)) {
            const configFileContent = fs.readFileSync(selectedConfigFile, 'utf-8');
            wasInstalledBefore = this.containsMarkers(configFileContent);
            if (!wasInstalledBefore) {
                const tabCompletionLoaderScript = this.getTabCompletionLoaderScript(configFileContent, selectedConfigFile);
                success.push(this.addConfig(selectedConfigFile, tabCompletionLoaderScript));
            }
        }
        this.logResult(success.some(Boolean), wasInstalledBefore);
    }
    containsMarkers(content) {
        return (content.includes(this.markers.SHELL_AND_PROFILE_START) &&
            content.includes(this.markers.SHELL_AND_PROFILE_END));
    }
    stripMarkersAndContent(startMarker, endMarker, content) {
        if (this.containsMarkers(content)) {
            const startIndex = content.indexOf(startMarker);
            const endIndex = content.indexOf(endMarker);
            if (startIndex !== -1 && endIndex !== -1 && endIndex > startIndex) {
                const beforeMarker = content.slice(0, startIndex);
                const afterMarker = content.slice(endIndex + endMarker.length);
                const removedContent = content.slice(startIndex, endIndex + endMarker.length);
                return {
                    updatedContent: (beforeMarker + afterMarker).trim(),
                    removedContent: removedContent.trim(),
                };
            }
        }
        return {
            updatedContent: content,
            removedContent: '',
        };
    }
    logResult(success, wasInstalledBefore) {
        var _a, _b, _c, _d;
        if (success) {
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-script-configured'));
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.warn(this.i18nInstance.t('warn-intellij-issue'));
            (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(this.i18nInstance.t('info-script-reopen-terminal'));
        }
        else if (!wasInstalledBefore) {
            (_d = this.logger) === null || _d === void 0 ? void 0 : _d.warn(this.i18nInstance.t('warn-script-install'));
        }
    }
    canWriteToConfig(configFilePath) {
        try {
            fs.accessSync(configFilePath, fs.constants.R_OK | fs.constants.W_OK);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    addConfig(configFilePath, newContent) {
        var _a, _b;
        try {
            fs.appendFileSync(configFilePath, newContent);
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-script-added', {
                path: configFilePath,
            }));
            return true;
        }
        catch (error) {
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.warn(this.i18nInstance.t('warn-script-added', {
                path: configFilePath,
                errorMsg: error.message,
            }));
            return false;
        }
    }
    resetCache() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const tempDir = path.join(os.tmpdir(), 'mgnl-v5');
            if (!fs.existsSync(tempDir))
                return;
            const items = yield fs.readdir(tempDir);
            const cacheFiles = items.filter((item) => item.startsWith('tab_completion_cache_'));
            let counter = 0;
            for (const file of cacheFiles) {
                const filePath = path.join(tempDir, file);
                try {
                    counter += 1;
                    fs.removeSync(filePath);
                }
                catch (_c) {
                    /* empty */
                }
            }
            if (counter > 0) {
                (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-removed-cache', {
                    count: counter,
                    path: tempDir,
                }));
            }
            else {
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(this.i18nInstance.t('info-removed-no-cache', { path: tempDir }));
            }
        });
    }
    uninstall() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f;
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(this.i18nInstance.t('info-uninstall'));
            if (fs.existsSync(this.paths.TAB_COMP_SCRIPT_PATH)) {
                fs.removeSync(this.paths.TAB_COMP_SCRIPT_PATH);
                (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(this.i18nInstance.t('info-removed-script', {
                    path: this.paths.MGNL_V5_DIR_PATH,
                }));
            }
            else {
                (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(this.i18nInstance.t('info-script-not-found', {
                    path: this.paths.MGNL_V5_DIR_PATH,
                }));
            }
            for (const script of this.systemScriptsPaths) {
                if (fs.existsSync(script)) {
                    const systemScriptContent = fs.readFileSync(script).toString();
                    const { updatedContent, removedContent } = this.stripMarkersAndContent(this.markers.SHELL_AND_PROFILE_START, this.markers.SHELL_AND_PROFILE_END, systemScriptContent);
                    if (removedContent.trim() === '') {
                        (_d = this.logger) === null || _d === void 0 ? void 0 : _d.info(this.i18nInstance.t('info-loader-script-not-found', {
                            path: script,
                        }));
                        return;
                    }
                    (_e = this.logger) === null || _e === void 0 ? void 0 : _e.info(this.i18nInstance.t('info-removal-from-script', {
                        path: script,
                    }));
                    // eslint-disable-next-line no-console
                    console.log('\n' + removedContent + '\n');
                    const { confirm } = yield inquirer.prompt({
                        type: 'confirm',
                        name: 'confirm',
                        message: this.i18nInstance.t('inquirer-confirm-script-removal'),
                        default: true,
                    });
                    if (confirm) {
                        fs.writeFileSync(script, updatedContent);
                    }
                    else {
                        (_f = this.logger) === null || _f === void 0 ? void 0 : _f.warn(this.i18nInstance.t('warn-not-confirmed-removal', {
                            path: script,
                        }));
                    }
                }
            }
        });
    }
}
