// @magnolia/cli-create-app-plugin/dist/create-app-plugin.js
import path from 'path';
import { createRequire } from 'module';
import { fileURLToPath } from 'url';
import { Option } from 'commander';
import ContentHelperPlugin, { cliPrototypesPath, } from '@magnolia/cli-content-helper';
import { initI18n } from '@magnolia/cli-helper/i18n';
import { findPrototypesPath, getAvailablePrototypes, } from '@magnolia/cli-helper/prototype';
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
const __dirname = path.dirname(fileURLToPath(import.meta.url));
export let i18nInstance;
export default class CreateAppPlugin extends ContentHelperPlugin {
    constructor(args = {}) {
        super(args, 'app');
        this.name = 'create-app';
        this.version = pkg.version;
        this.usage = '<name> [options]' +
            '\nParameter: <name> The name of the app (recommended plural).';
        i18nInstance = initI18n(this.name, 'translation', path.join(__dirname, 'lib/locales'));
        const prototypesPath = findPrototypesPath(args === null || args === void 0 ? void 0 : args.prototypeDir, cliPrototypesPath);
        const prototypes = getAvailablePrototypes(prototypesPath);
        const cmdOptionPrototype = prototypes
            ? 'option-prototype-description.with-prototypes'
            : 'option-prototype-description.no-prototypes';
        this.description = i18nInstance.t('description');
        this.options = [
            new Option('-ct, --content-type <name>', i18nInstance.t('option-contentType-description')),
            new Option('-lm, --light-module [name]', i18nInstance.t('option-lightModule-description')),
            new Option('-pd, --prototype-dir <path>', i18nInstance.t('option-prototypeDir-description')),
            new Option('-p, --prototype [name]', i18nInstance.t(cmdOptionPrototype, {
                prototypes,
                path: prototypesPath,
            })),
        ];
    }
}
