// @magnolia/cli-template-helper/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import path from 'path';
import fs from 'fs-extra';
import { i18nInstance, logger } from '../cli-template-helper.js';
import inquirer from 'inquirer';
import { CreateError, prependNumbersToChoices, } from '@magnolia/cli-helper/general-utils';
import { promptForDirOrFile } from '@magnolia/cli-helper/prompt';
import { validateName } from '@magnolia/cli-helper';
// @ts-ignore
import isVarName from 'is-valid-var-name';
const getPluginName = (context) => {
    return `Create${context === 'pages' ? 'Page' : 'Component'}Plugin`;
};
const getDefaultComponentsMappingFileContent = (mapping) => {
    const defaultComponentMappingContent = '\n// Please visit documentation:\n' +
        '// https://docs.magnolia-cms.com/headless/spa-development/mapping-between-magnolia-and-spa-components/\n' +
        '// to learn how to map between Magnolia and SPA components\n' +
        '\nconst componentMappings = {\n' +
        "\t'componentMappings': {\n" +
        (mapping ? '\t\t' + mapping + '\n' : '\t\t\n') +
        '\t}\n' +
        '};\n';
    return defaultComponentMappingContent;
};
const pluginCwd = process.env.INIT_CWD || process.cwd();
function getEmptyPrefix(str) {
    const modifiedStr = str.replace(/^\n+/, '');
    const match = modifiedStr.match(/^([\s\t]*)/);
    return match ? match[1] : '';
}
function getHalfEmptyPrefix(str) {
    return str.substring(0, str.length / 2);
}
function getDestLightModuleName(lightModulePath) {
    if (!fs.existsSync(lightModulePath)) {
        throw new CreateError(i18nInstance.t('error-lm-does-not-exist', {
            lightModulePath: lightModulePath,
        }));
    }
    return path.basename(lightModulePath);
}
function getAllPaths(basePath) {
    const filePaths = [];
    function readDirSync(currentPath) {
        const items = fs.readdirSync(currentPath, { withFileTypes: true });
        items.forEach((item) => {
            const fullPath = path.join(currentPath, item.name);
            if (item.isFile()) {
                filePaths.push(fullPath);
            }
            else if (item.isDirectory()) {
                readDirSync(fullPath);
            }
        });
    }
    readDirSync(basePath);
    return filePaths;
}
function promptForPrototype(pathToFrameworkContext, framework) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!fs.existsSync(pathToFrameworkContext)) {
            throw new CreateError(i18nInstance.t('error-path-to-fw-context-does-not-exist', {
                framework: framework,
            }));
        }
        const contextContent = fs.readdirSync(pathToFrameworkContext);
        const availablePrototypes = contextContent.filter((entry) => {
            const fullPath = path.join(pathToFrameworkContext, entry);
            return fs.statSync(fullPath).isDirectory();
        });
        if (availablePrototypes.length === 0) {
            throw new CreateError(i18nInstance.t('error-no-prototypes-available'));
        }
        const { selectedPrototype } = yield inquirer.prompt([
            {
                type: 'list',
                name: 'selectedPrototype',
                message: i18nInstance.t('inquirer-prompt-choose-prototype', {
                    framework: framework,
                }),
                choices: prependNumbersToChoices(availablePrototypes),
            },
        ]);
        return selectedPrototype;
    });
}
function promptForType(availableTypes, prototype) {
    return __awaiter(this, void 0, void 0, function* () {
        const { selectedType } = yield inquirer.prompt([
            {
                type: 'rawlist',
                name: 'selectedType',
                message: i18nInstance.t('inquirer-prompt-choose-type-of-prototype', { prototype: prototype }),
                choices: prependNumbersToChoices(availableTypes),
            },
        ]);
        return selectedType;
    });
}
function promptForCreationOfNotExistingResource(msg_1, resourcePath_1, type_1) {
    return __awaiter(this, arguments, void 0, function* (msg, resourcePath, type, newFileContent = undefined) {
        const answer = yield inquirer.prompt([
            {
                type: 'confirm',
                name: 'confirm',
                message: msg,
                default: true,
            },
        ]);
        if (answer.confirm) {
            if (type == 'dir') {
                fs.ensureDirSync(resourcePath);
                logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-folder-created', {
                    name: path.basename(resourcePath),
                    path: resourcePath,
                }));
            }
            else {
                fs.ensureFileSync(resourcePath);
                if (newFileContent !== undefined) {
                    fs.writeFileSync(resourcePath, newFileContent);
                }
                logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-file-created', {
                    name: path.basename(resourcePath),
                    path: resourcePath,
                }));
            }
        }
        return answer.confirm;
    });
}
function promptForComponentsMappingFile() {
    return __awaiter(this, void 0, void 0, function* () {
        const answer = yield inquirer.prompt([
            {
                type: 'confirm',
                name: 'confirm',
                message: i18nInstance.t('inquirer-prompt-confirm-cmp-file'),
                default: true,
            },
        ]);
        if (answer.confirm) {
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('inquirer-prompt-choose-components-mapping-file'));
            return yield promptForDirOrFile({
                type: 'file',
                logger,
                createNew: true,
                fileContent: getDefaultComponentsMappingFileContent(),
            });
        }
        return '';
    });
}
function getAbsolutePath(relativePath) {
    if (relativePath === undefined || relativePath === '')
        return relativePath;
    return path.normalize(path.resolve(process.env.INIT_CWD || process.cwd(), relativePath));
}
function convertToPascalCase(name) {
    if (!name.includes('-')) {
        return name;
    }
    return name
        .split('-')
        .map((word, index) => {
        if (index === 0 && /^[a-z]/.test(word)) {
            return word.charAt(0).toUpperCase() + word.slice(1);
        }
        return word.charAt(0).toUpperCase() + word.slice(1);
    })
        .join('');
}
function promptForComponentName(componentType) {
    return __awaiter(this, void 0, void 0, function* () {
        while (true) {
            const { componentNameInput } = yield inquirer.prompt([
                {
                    type: 'input',
                    name: 'componentNameInput',
                    message: i18nInstance.t('inquirer-prompt-name', {
                        type: componentType,
                    }),
                },
            ]);
            try {
                return parseName(componentNameInput, componentType);
            }
            catch (_a) {
                logger.error(i18nInstance.t('error-name-invalid-js-var', {
                    name: componentNameInput,
                    type: componentType,
                }));
            }
        }
    });
}
function parseName(name, componentType) {
    const lastSlashIndex = name.lastIndexOf('/');
    const subDirs = name.substring(0, lastSlashIndex);
    const originalName = name.substring(lastSlashIndex + 1);
    const componentName = helper.validateAndResolveComponentName(originalName, componentType);
    return { originalName, componentName, subDirs };
}
function validateAndResolveComponentName(name, componentType) {
    if (!name) {
        throw new CreateError(i18nInstance.t('error-name-required', { type: componentType }));
    }
    validateName(componentType, name);
    const componentName = convertToPascalCase(name);
    if (!isVarName(componentName)) {
        throw new CreateError(i18nInstance.t('error-name-invalid-js-var', {
            type: componentType,
            name,
        }));
    }
    return componentName;
}
export const helper = {
    getAbsolutePath,
    pluginCwd,
    getEmptyPrefix,
    getHalfEmptyPrefix,
    getDestLightModuleName,
    getAllPaths,
    promptForPrototype,
    promptForType,
    promptForCreationOfNotExistingResource,
    promptForComponentsMappingFile,
    getDefaultComponentsMappingFileContent,
    getPluginName,
    validateAndResolveComponentName,
    parseName,
    promptForComponentName,
};
