// @magnolia/cli-start-plugin/dist/start-plugin.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import path from 'path';
import { fileURLToPath } from 'url';
import { PluginTemplate } from '@magnolia/cli-plugin-template';
import { Option } from 'commander';
import { createRequire } from 'module';
import { helper } from './lib/helper.js';
import { mgnlStart } from './lib/mgnl-start.js';
import { tpHelper } from './lib/tomcatPathHelper.js';
import { resolveLightModulesPath, getMgnlConfig, getRelativePathToMGNLConfig, initI18n, modifyConfig, checkFlagsValue, } from '@magnolia/cli-helper';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
export let logger;
export let i18nInstance = {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    t(key, options) {
        return key;
    },
};
export default class StartPlugin extends PluginTemplate {
    constructor(args) {
        super();
        this.pluginName = this.constructor.name;
        this.name = 'start';
        this.version = pkg.version;
        this.usage = '[options]';
        this.configUpdates = {
            sharedProps: {},
            pluginProps: {},
        };
        i18nInstance = initI18n(this.name, 'translation', path.join(__dirname, 'lib/locales'));
        this.description = i18nInstance.t('description');
        this.options = [
            new Option('-t, --tomcat [path]', i18nInstance.t('option-tomcat-description')),
            new Option('-lmp, --light-modules-path [path]', i18nInstance.t('option-lightModulesPath-description')),
        ];
        this.pluginArgs = args !== null && args !== void 0 ? args : {};
    }
    init(winstonLogger) {
        return __awaiter(this, void 0, void 0, function* () {
            logger = winstonLogger;
        });
    }
    validateAndResolveArgs(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            checkFlagsValue(this.options, options);
            const config = yield getMgnlConfig(logger);
            let lightModulesPath;
            if (options.lightModulesPath || (config === null || config === void 0 ? void 0 : config.lightModulesPath)) {
                const lmp = yield resolveLightModulesPath(options, config, logger, true);
                lightModulesPath = lmp.lightModulesPath;
                if (lmp.newLightModulesPath) {
                    this.configUpdates.sharedProps.lightModulesPath =
                        lmp.newLightModulesPath;
                }
            }
            const { tomcatPath: apacheTomcatFolderPath, updated: tomcatPathUpdated, } = yield tpHelper.getTomcatPath(options, this.pluginArgs);
            if (!apacheTomcatFolderPath) {
                logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-tomcat-not-found', {
                    dir: (_a = apacheTomcatFolderPath !== null && apacheTomcatFolderPath !== void 0 ? apacheTomcatFolderPath : options.tomcat) !== null && _a !== void 0 ? _a : helper.getCwd(),
                }));
                logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-unable-to-start-mgnl'));
                helper.terminateProcessWithErrorMessage(i18nInstance.t('error-use-jumpstart'), 1);
            }
            else if (tomcatPathUpdated) {
                this.configUpdates.pluginProps.tomcatPath =
                    getRelativePathToMGNLConfig(apacheTomcatFolderPath);
            }
            return {
                apacheTomcatFolderPath,
                lightModulesPath,
            };
        });
    }
    start(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const isJavaSetCorrectly = yield helper.checkJavaEnvironment();
            if (!isJavaSetCorrectly) {
                return;
            }
            const { apacheTomcatFolderPath, lightModulesPath } = yield this.validateAndResolveArgs(options);
            yield modifyConfig(this.configUpdates, this.pluginName, logger);
            this.chosenProject = mgnlStart.getTomcat(apacheTomcatFolderPath, lightModulesPath);
            yield ((_a = this.chosenProject) === null || _a === void 0 ? void 0 : _a.start());
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            yield ((_a = this.chosenProject) === null || _a === void 0 ? void 0 : _a.stop());
            process.exit(0);
        });
    }
}
