// @magnolia/cli-start-plugin/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import fs from 'fs-extra';
import path from 'path';
import { i18nInstance, logger } from '../start-plugin.js';
import { execa } from 'execa';
function getCwd() {
    var _a;
    return (_a = process.env.INIT_CWD) !== null && _a !== void 0 ? _a : process.cwd();
}
function locateTomcatDirectory(startPath) {
    if (!fs.existsSync(startPath)) {
        return;
    }
    const isApacheTomcat = (pathPath) => {
        return /apache-tomcat*/.test(pathPath);
    };
    if (isApacheTomcat(path.basename(startPath))) {
        return startPath;
    }
    return fs
        .readdirSync(startPath, { withFileTypes: true })
        .filter((it) => it.isDirectory() && isApacheTomcat(it.name))
        .map((it) => (it !== undefined ? path.join(startPath, it.name) : it))
        .find((it) => it !== undefined);
}
function terminateProcess(code, tailProcess) {
    if (tailProcess !== undefined) {
        tailProcess.kill();
    }
    process.exit(code);
}
function terminateProcessWithErrorMessage(msg, exitCode) {
    logger === null || logger === void 0 ? void 0 : logger.error(msg);
    process.exit(exitCode);
}
const validatePath = (currentPath, context, source) => __awaiter(void 0, void 0, void 0, function* () {
    if (!fs.existsSync(currentPath)) {
        let notExistMessage = 'warn-provided-path-does-not-exist';
        if (source === 'option') {
            notExistMessage = 'warn--path-does-not-exist-option';
        }
        else if (source === 'plugin') {
            notExistMessage = 'warn--path-does-not-exist-plugin';
        }
        else if (source === 'global') {
            notExistMessage = 'warn--path-does-not-exist-global';
        }
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t(notExistMessage, {
            context: context === 'lm' ? 'light modules' : 'apache-tomcat',
            currentPath: currentPath,
        }));
        return undefined;
    }
    return currentPath;
});
const getAbsolutePath = (relativePath) => {
    if (relativePath === undefined) {
        return relativePath;
    }
    return path.resolve(helper.getCwd(), relativePath);
};
function isJavaInstalled() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const { stdout, stderr } = yield execa('java', ['-version'], {
                cwd: process.cwd(),
            });
            const output = stdout + stderr;
            if (output.toLowerCase().includes('java') ||
                output.toLowerCase().includes('openjdk')) {
                return true;
            }
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
        }
        catch (error) {
            return false;
        }
        return false;
    });
}
function checkJavaEnvironment() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!(yield isJavaInstalled())) {
            logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance === null || i18nInstance === void 0 ? void 0 : i18nInstance.t('error-java-not-installed'));
            return false;
        }
        if (!process.env.JAVA_HOME) {
            logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance === null || i18nInstance === void 0 ? void 0 : i18nInstance.t('error-java-home-not-set'));
            return false;
        }
        return true;
    });
}
export const helper = {
    getCwd,
    terminateProcess,
    terminateProcessWithErrorMessage,
    locateTomcatDirectory,
    validatePath,
    getAbsolutePath,
    checkJavaEnvironment,
};
