// @magnolia/cli-start-plugin/dist/lib/mgnl-start.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { helper } from './helper.js';
import { i18nInstance, logger } from '../start-plugin.js';
import fs from 'fs-extra';
import { execFile } from 'child_process';
import path from 'path';
import os from 'os';
import { Tail } from 'tail';
import net from 'net';
import * as xml2js from 'xml2js';
import ora from 'ora';
function getTomcat(apacheTomcatFolderPath, lightModulesPath) {
    const tomcatProject = {
        tomcatPath: apacheTomcatFolderPath,
        lightModulesPath,
        childProcess: undefined,
        start: () => __awaiter(this, void 0, void 0, function* () { return yield mgnlStart.startUpMagnolia(tomcatProject); }),
        stop: () => __awaiter(this, void 0, void 0, function* () { return yield mgnlStart.gracefulShutdownMagnolia(tomcatProject); }),
    };
    return tomcatProject;
}
function startUpMagnolia(project) {
    return __awaiter(this, void 0, void 0, function* () {
        yield checkPort(project.tomcatPath);
        startMagnoliaControl(project);
        waitForFileCreation(project.tomcatPath);
    });
}
function gracefulShutdownMagnolia(project) {
    return __awaiter(this, void 0, void 0, function* () {
        const shutdownScript = path.join(project.tomcatPath, 'bin', os.platform() === 'win32' ? 'shutdown.bat' : 'magnolia_control.sh');
        const executeShutdown = () => __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                execFile(shutdownScript, ['stop'], { cwd: path.join(project.tomcatPath, 'bin'), shell: true }, (error, stdout, stderr) => {
                    if (error || (stderr === null || stderr === void 0 ? void 0 : stderr.includes('SEVERE:'))) {
                        reject(new Error(stderr || (error === null || error === void 0 ? void 0 : error.message)));
                    }
                    else {
                        resolve({ stdout: stdout, stderr: stderr });
                    }
                });
            });
        });
        const spinner = ora().start(i18nInstance.t('ora-start-shutdown-attempt'));
        const attemptTotal = 10; // 10 tries
        const attemptRetryDelay = 10000; // 10 seconds
        for (let attempt = 0; attempt < attemptTotal; attempt++) {
            try {
                const output = yield executeShutdown();
                // eslint-disable-next-line no-console
                console.log(output.stdout);
                // eslint-disable-next-line no-console
                console.error(output.stderr);
                // Wait 5 seconds so tomcat has time to disconnect properly
                yield new Promise((resolve) => setTimeout(resolve, 5000));
                spinner.succeed(i18nInstance.t('ora-succeed-shutdown'));
                if (project.childProcess) {
                    helper.terminateProcess(0, project.childProcess);
                }
                return;
            }
            catch (error) {
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-an-error-during-shutdown'));
                // eslint-disable-next-line no-console
                console.error(error.message);
                spinner.fail(i18nInstance.t('ora-fail-attempt', { attempt: attempt + 1 }));
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-may-be-starting-during-shutdown'));
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-during-shutdown-wait-logs-will-appear'));
                if (attempt < 9) {
                    spinner.start(i18nInstance.t('ora-start-waiting-s', {
                        seconds: attemptRetryDelay / 1000,
                        attempt: attemptTotal - 1 - attempt,
                    }));
                    yield new Promise((resolve) => setTimeout(resolve, attemptRetryDelay));
                    spinner.stop();
                    spinner.start(i18nInstance.t('ora-start-shutdown-retry', {
                        attempt: attempt + 2,
                        total: attemptTotal,
                    }));
                }
                else {
                    spinner.fail(i18nInstance.t('ora-fail-all-shutdown-attempts', {
                        total: attemptRetryDelay,
                    }));
                    logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-mgnl-shutdown-failed', {
                        file: shutdownScript,
                    }));
                    if (project.childProcess) {
                        helper.terminateProcess(1, project.childProcess);
                    }
                    throw error;
                }
            }
        }
    });
}
const startMagnoliaControl = (project) => {
    const magnoliaControl = path.join(project.tomcatPath, 'bin', os.platform() === 'win32'
        ? 'magnolia_control.bat'
        : 'magnolia_control.sh');
    if (!fs.existsSync(magnoliaControl)) {
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-mgnl-control-file-missing'));
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-unable-to-start-mgnl'));
        return;
    }
    const args = ['start', '--ignore-open-files-limit'];
    const catalinaOpts = [
        process.env.CATALINA_OPTS,
        project.lightModulesPath === undefined
            ? ''
            : `-Dmagnolia.resources.dir="${project.lightModulesPath}"`,
    ]
        .filter(Boolean)
        .join(' ');
    const execOptions = {
        cwd: path.join(project.tomcatPath, 'bin'),
        env: Object.assign(Object.assign({}, process.env), { CATALINA_HOME: project.tomcatPath, CATALINA_OPTS: catalinaOpts }),
        shell: true,
    };
    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-starting-magnolia', {
        apacheTomcatFolder: project.tomcatPath,
    }));
    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-press-ctrl-c'));
    const child = execFile(magnoliaControl, args, execOptions, (error) => {
        if (error) {
            logger.error(error.message);
            helper.terminateProcess(1, child);
        }
    });
    project.childProcess = child;
};
const startTail = (file) => {
    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-displaying-log', { file }));
    const tail = new Tail(file);
    tail.on('line', function (data) {
        // eslint-disable-next-line no-console
        console.log(data);
    });
    tail.on('error', function (error) {
        // eslint-disable-next-line no-console
        console.error(error);
    });
};
const waitForFileCreation = (projectPath) => {
    const file = path.join(projectPath, 'logs', 'catalina.out');
    if (fs.existsSync(file)) {
        startTail(file);
        return;
    }
    const logsPath = path.join(projectPath, 'logs');
    fs.ensureDirSync(logsPath);
    const watcher = fs.watch(logsPath, (eventType, filename) => {
        if (filename !== null) {
            if (filename === path.basename(file) && fs.existsSync(file)) {
                watcher.close();
                startTail(file);
            }
        }
    });
};
const checkPort = (projectPath) => __awaiter(void 0, void 0, void 0, function* () {
    const host = '127.0.0.1';
    let port = 8080;
    const serverXml = path.join(projectPath, 'conf', 'server.xml');
    if (!fs.existsSync(serverXml)) {
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-missing-server-xml-file', {
            serverXml,
        }));
        helper.terminateProcessWithErrorMessage(i18nInstance.t('error-unable-to-start-mgnl'), 1);
    }
    xml2js.parseString(fs.readFileSync(serverXml), (err, result) => {
        var _a, _b, _c, _d, _e, _f;
        if (err != null) {
            logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-cannot-parse-server-xml-port'));
        }
        else {
            const parsedPort = (_f = (_e = (_d = (_c = (_b = (_a = result.Server) === null || _a === void 0 ? void 0 : _a.Service) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.Connector) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.$) === null || _f === void 0 ? void 0 : _f.port;
            if (parsedPort === undefined) {
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-cannot-parse-server-xml-port'));
            }
            else {
                port = parsedPort;
            }
        }
    });
    const isFree = yield new Promise((resolve) => {
        const connection = net
            .createConnection({ port, host })
            .once('connect', () => {
            connection.end();
            resolve(false);
        })
            .once('error', () => {
            resolve(true);
        });
    });
    if (isFree === false) {
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-port-in-use', { port }));
        helper.terminateProcessWithErrorMessage(i18nInstance.t('error-unable-to-start-mgnl'), 1);
    }
});
export const mgnlStart = {
    getTomcat,
    startUpMagnolia,
    gracefulShutdownMagnolia,
    startTail,
    waitForFileCreation,
};
