// @magnolia/cli-generate-plugin/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import i18next from 'i18next';
import Backend from 'i18next-fs-backend';
import inquirer from 'inquirer';
import { i18nInstance, logger } from '../generate-plugin.js';
import path from 'path';
import fs from 'fs-extra';
import { fileURLToPath } from 'url';
export class MgnlCliError extends Error {
}
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
export function initI18n(pluginName, namespace = 'translation') {
    const lng = 'en';
    const localesPath = path.join(__dirname, `locales/${lng}`);
    const namespaces = fs
        .readdirSync(localesPath)
        .filter((file) => file.endsWith('.json'))
        .map((file) => file.replace('.json', ''));
    if (!namespaces.includes(namespace)) {
        // eslint-disable-next-line no-console
        console.warn(`${pluginName}: Requested translation file "${namespace}" not found at "${localesPath}".\n`);
        if (namespaces.length > 0) {
            // eslint-disable-next-line no-console
            console.warn(`Using "${namespaces[0]}" instead.\n`);
        }
        namespace = namespaces[0];
    }
    const newI18nInstance = i18next.createInstance();
    newI18nInstance.use(Backend).init({
        joinArrays: '\n',
        lng: lng,
        fallbackLng: 'en',
        ns: namespaces,
        defaultNS: namespace,
        initImmediate: false,
        interpolation: {
            escapeValue: false,
        },
        backend: {
            loadPath: path.join(__dirname, 'locales/{{lng}}/{{ns}}.json'),
        },
    });
    return newI18nInstance;
}
export function resolveOptionOrPrompt(promptMsg, defaultPromptValue, isValid, optionValue) {
    return __awaiter(this, void 0, void 0, function* () {
        return ((yield validateOption(optionValue, isValid)) ||
            (yield promptForValue(promptMsg, defaultPromptValue, isValid)));
    });
}
export function promptForValue(message_1) {
    return __awaiter(this, arguments, void 0, function* (message, defaultValue = null, isValid) {
        let value = yield getUserInput(message, defaultValue);
        if (isValid === undefined) {
            return value;
        }
        while (!isValid(value)) {
            value = yield promptForValue(message, defaultValue, isValid);
        }
        return value;
    });
}
export function validateOption(value, isValid) {
    return __awaiter(this, void 0, void 0, function* () {
        if (value === undefined) {
            return undefined;
        }
        if (isValid === undefined) {
            return value;
        }
        if (!isValid(value)) {
            throw new MgnlCliError(i18nInstance.t('error-invalid-value', { value }));
        }
        return value;
    });
}
function getUserInput(message_1) {
    return __awaiter(this, arguments, void 0, function* (message, defaultValue = null) {
        const promptInput = yield inquirer.prompt([
            {
                type: 'input',
                name: 'value',
                message,
                default: defaultValue !== null && defaultValue !== void 0 ? defaultValue : undefined,
            },
        ]);
        return promptInput.value.trim();
    });
}
export function isValidClassName(className) {
    if (className.length > 214 || className.length === 0) {
        return false;
    }
    const validClassNameRegex = /^[A-Za-z][A-Za-z0-9_$]*$/;
    if (!validClassNameRegex.test(className)) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-invalid-class-name', { className }));
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-class-name-validation'));
        return false;
    }
    return true;
}
export function isValidPluginName(pluginName) {
    // Some of the rules taken from: https://docs.npmjs.com/cli/v6/configuring-npm/package-json#name
    if (pluginName.length > 214) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-plugin-name-validation'));
        return false;
    }
    // The pluginName is also used as a file name therefore it is more restrictive
    const packageNameRegex = /^[a-z0-9][a-z0-9-_]*$/;
    if (!packageNameRegex.test(pluginName)) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-invalid-plugin-name', { pluginName }));
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-plugin-name'));
        return false;
    }
    return true;
}
