// @magnolia/cli-jumpstart-plugin/dist/lib/extensions.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import fs from 'fs-extra';
import path from 'path';
import inquirer from 'inquirer';
import { execa } from 'execa';
import { parse } from 'yaml';
import { i18nInstance, logger } from '../jumpstart-plugin.js';
import ora from 'ora';
import { determinePackageManager } from './helper.js';
import { pathToFileURL } from 'url';
import { globSync } from 'glob';
import _ from 'underscore';
function pause(milliSeconds) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            }, milliSeconds);
        });
    });
}
function callExtensionFunction(func_1, context_1, name_1) {
    return __awaiter(this, arguments, void 0, function* (func, context, name, delay = 200, retry = 0, maxRetry = 0) {
        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-extensions-calling-fn', {
            name: name,
        }));
        try {
            yield func(context);
        }
        catch (error) {
            if (retry === maxRetry) {
                throw error;
            }
            logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-extensions-calling-fn-retry', {
                name: name,
                retry: retry,
                maxRetry: maxRetry,
                delay: delay,
            }));
            yield pause(delay);
            yield callExtensionFunction(func, context, name, delay, retry + 1, maxRetry);
        }
    });
}
function executePostCommands() {
    return __awaiter(this, arguments, void 0, function* (cmds = [], extensions, context) {
        for (const cmd of cmds) {
            const func = extensions[cmd.function];
            if (func) {
                yield callExtensionFunction(func, context, cmd.function, cmd.delay, 0, cmd.retry);
            }
            else {
                logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-extensions-fn-not-found', {
                    function: cmd.function,
                }));
            }
        }
    });
}
// evaluate the given yaml file and create cli prompts out of it, return the answers
export const evaluateCustomPrompts = (extensionsPath) => __awaiter(void 0, void 0, void 0, function* () {
    if (!fs.existsSync(extensionsPath)) {
        return;
    }
    const doc = parse(fs.readFileSync(extensionsPath, 'utf8'));
    const baseDir = path.parse(extensionsPath).dir;
    const packageJson = path.join(baseDir, 'package.json');
    if (fs.existsSync(packageJson)) {
        const npmISpinner = ora().start(i18nInstance.t('ora-start-installing-ext-dep'));
        const packageManager = determinePackageManager(baseDir);
        try {
            yield execa(packageManager, ['install'], {
                buffer: true,
                cwd: baseDir,
            });
            npmISpinner.succeed(i18nInstance.t('ora-succeed-extensions-install-done'));
        }
        catch (e) {
            npmISpinner.fail(i18nInstance.t('ora-fail-pm-install', {
                packageManager: packageManager,
                errorMsg: e.message,
            }));
            return;
        }
    }
    const extensions = doc.extend
        ? yield import(pathToFileURL(path.join(baseDir, doc.extend)).href)
        : null;
    return {
        answers: doc.prompts ? yield inquirer.prompt(doc.prompts) : {},
        executePostCommands: (context) => __awaiter(void 0, void 0, void 0, function* () { var _a; return yield executePostCommands((_a = doc === null || doc === void 0 ? void 0 : doc.commands) === null || _a === void 0 ? void 0 : _a.post, extensions, context); }),
    };
});
export const compileCustomPrompts = (answers_1, ...args_1) => __awaiter(void 0, [answers_1, ...args_1], void 0, function* (answers, pattern = '**/**/**/*.{yaml,tsx,ts,js,jsx,vue,md,json}') {
    const init = globSync(pattern, { ignore: 'node_modules/**' }).map((itm) => path.join(process.cwd(), itm));
    const pkg = JSON.parse(fs.readFileSync(path.join(process.cwd(), './package.json'), 'utf8'));
    pkg.answers = answers;
    for (const item of init) {
        const file = fs.readFileSync(item, 'utf8');
        const doc = _.template(file)(pkg);
        fs.outputFileSync(item, doc, 'utf8');
    }
});
