// @magnolia/cli-jumpstart-plugin/dist/lib/handleMicroprofileConfig.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import fs from 'fs';
import path from 'path';
import { getMicroprofileConfigLocation } from './helper.js';
import { i18nInstance, logger } from '../jumpstart-plugin.js';
export const handleMicroprofileConfig = (bundle, dest) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b;
    if (!((_a = bundle.version) === null || _a === void 0 ? void 0 : _a.startsWith('6.3')) &&
        !((_b = bundle.version) === null || _b === void 0 ? void 0 : _b.startsWith('6.4')))
        return;
    const tomcatPath = path.join(path.resolve(dest), '../../');
    const webappName = path.basename(dest);
    const pathToMicroprofileConfig = getMicroprofileConfigLocation(tomcatPath, webappName);
    processMicroprofileConfigFile(pathToMicroprofileConfig);
});
export function convertObjectToProperties(obj) {
    const props = [];
    Object.keys(obj).forEach((key) => {
        const val = obj[key];
        if (val && typeof val === 'object' && !Array.isArray(val)) {
            const children = convertObjectToProperties(val);
            for (const child of children) {
                props.push(key + '.' + child);
            }
        }
        else {
            props.push(key + '=' + val);
        }
    });
    return props;
}
function createMicroprofileConfigFile(file) {
    const microprofileConfig = {
        magnolia: {
            publishing: {
                'receivers[0]': {
                    name: 'magnoliaPublic8080',
                    url: 'http://localhost:8080/magnoliaPublic',
                    enabled: true,
                },
            },
        },
    };
    fs.writeFileSync(file, convertObjectToProperties(microprofileConfig).join('\n'), 'utf8');
    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-created-microprofile', { file }));
}
function processMicroprofileConfigFile(file) {
    if (fs.existsSync(file)) {
        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-microprofile-exists'));
    }
    else {
        createMicroprofileConfigFile(file);
    }
}
