// @magnolia/cli-jumpstart-plugin/dist/lib/helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { handleMGNLConfigFile } from './config-helper.js';
import { handlePackageJSON } from './pj-helper.js';
import { PostCommands } from '../types/types.js';
import ora from 'ora';
import { execa } from 'execa';
import path from 'path';
import fs from 'fs-extra';
import { findExtractedApacheTomcatDir } from './extract.js';
import walk from 'walk';
import { i18nInstance, logger } from '../jumpstart-plugin.js';
import inquirer from 'inquirer';
import { CreateError } from '@magnolia/cli-helper/general-utils';
import { getRelativePathToMGNLConfig } from '@magnolia/cli-helper';
import chalk from 'chalk';
export const handleLightModulesFolder = () => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line no-console
    console.log(chalk.cyanBright('==========================================='));
    let lightModulesPath;
    try {
        lightModulesPath = (yield findLightModulesFolder(process.cwd()));
        logger === null || logger === void 0 ? void 0 : logger.info(chalk.green.bold('✅ ' + i18nInstance.t('info-found-lm', { lightModulesPath })));
    }
    catch (_a) {
        lightModulesPath = path.join(process.cwd(), 'light-modules');
        try {
            fs.mkdirSync(lightModulesPath, { recursive: true });
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-lm-created', {
                lightModulesPath: lightModulesPath,
            }));
        }
        catch (error) {
            logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-lm-create-fail', {
                errorMsg: error.message,
            }));
            return;
        }
    }
    let apacheTomcatPath;
    try {
        apacheTomcatPath = yield findExtractedApacheTomcatDir(process.cwd());
    }
    catch (error) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-cannot-modify-webapp', {
            errorMsg: error.message,
        }));
        return;
    }
    const webApps = findWebAppsList(apacheTomcatPath);
    for (const webApp of webApps) {
        const lightModulesPathToMagnoliaHomeRelPath = process.platform === 'win32'
            ? path.relative(webApp, lightModulesPath).replace(/\\/g, '/')
            : path.relative(webApp, lightModulesPath);
        const props = {
            'magnolia.resources.dir': '${magnolia.home}/' + lightModulesPathToMagnoliaHomeRelPath,
            'magnolia.update.auto': 'true',
        };
        if (webApp.endsWith('magnoliaAuthor')) {
            props['magnolia.develop'] = 'true';
            props['magnolia.graphql.introspection.enabled'] = 'true';
            props['magnolia.personalization.rest.allowVariants'] = 'true';
        }
        const webAppPropertiesFilePath = path.join(webApp, 'WEB-INF', 'config', 'default', 'magnolia.properties');
        editProperties(webAppPropertiesFilePath, props);
    }
    // eslint-disable-next-line no-console
    console.log(chalk.cyanBright('==========================================='));
});
const editProperties = (filePath, props) => {
    if (!fs.existsSync(filePath)) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-webapp-props-not-exist', {
            filePath: filePath,
        }));
        return;
    }
    const propertiesFileContent = fs.readFileSync(filePath, 'utf-8');
    const replacedContent = replaceInConfig(propertiesFileContent, props);
    try {
        fs.writeFileSync(filePath, replacedContent);
        Object.keys(props).forEach((key) => {
            logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-webapp-prop-setting', {
                key: key,
                propsKey: props[key],
                filePath: filePath,
            }));
        });
    }
    catch (error) {
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-webapp-props-update-error', {
            errorMsg: error.message,
            path: filePath,
        }));
    }
};
const replaceInConfig = (config, props) => {
    Object.keys(props).forEach((key) => {
        const regexPattern = '(' + key.replace(/\./g, '\\.') + ')(\s*=\s*)(.+)'; // eslint-disable-line no-useless-escape
        const regex = new RegExp(regexPattern);
        const replacement = '$1$2' + props[key];
        if (regex.test(config)) {
            config = config.replace(regex, replacement);
        }
        else {
            config += `\n${key}=${props[key]}`;
        }
    });
    return config;
};
export const handleLightModulesPathInTemplate = (template) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const lightModulesPath = yield findLightModulesFolder(process.cwd()).catch(() => undefined);
    if (typeof lightModulesPath === 'string' &&
        lightModulesPath !== '' &&
        ((_a = template.configVars) === null || _a === void 0 ? void 0 : _a.lightModulesPath) === undefined) {
        template.configVars = Object.assign(Object.assign({}, template.configVars), { lightModulesPath: getRelativePathToMGNLConfig(lightModulesPath) });
    }
});
export const findLightModulesFolder = (p) => {
    return new Promise((resolve, reject) => {
        const walker = walk.walk(p, { followLinks: false });
        walker.on('directory', (root, { name }, next) => {
            if (name.match(/(light-modules)/)) {
                resolve(path.join(root, name));
            }
            next();
        });
        walker.on('end', function () {
            return reject(new CreateError(i18nInstance.t('error-lm-not-found')));
        });
    });
};
export const findWebAppsList = (apacheTomcatPath) => {
    const webApps = [];
    const webAppsFolder = path.join(apacheTomcatPath, 'webapps');
    const files = fs.readdirSync(webAppsFolder);
    for (const file of files) {
        const filePath = path.join(webAppsFolder, file);
        if (fs.statSync(filePath).isDirectory()) {
            webApps.push(filePath);
        }
    }
    return webApps;
};
export const initializeNodeProject = () => __awaiter(void 0, void 0, void 0, function* () {
    yield handlePackageJSON();
    yield handleMGNLConfigFile();
});
export const installCLI = () => __awaiter(void 0, void 0, void 0, function* () {
    return yield installPackage(determinePackageManager(), '@magnolia/cli@5', 'mgnl-cli');
});
export const installPackage = (packageManager, packageReference, packageName) => __awaiter(void 0, void 0, void 0, function* () {
    if (!packageName) {
        packageName = packageReference;
    }
    const spinner = ora().start(i18nInstance.t('ora-start-installing-package', {
        packageName: packageName,
    }));
    try {
        yield execa(packageManager, [packageManager === 'npm' ? 'install' : 'add', packageReference], {
            buffer: true,
            cwd: process.cwd(),
        });
        spinner.succeed(i18nInstance.t('ora-succeed-package-installed', {
            packageName: packageName,
        }));
        return true;
    }
    catch (error) {
        spinner.fail(i18nInstance.t('ora-fail-installing-package', {
            packageName: packageName,
            errorMsg: error.message,
        }));
        return false;
    }
});
export const determinePackageManager = (baseDirectory = process.cwd()) => {
    const yarnLockPath = path.join(baseDirectory, 'yarn.lock');
    const packageLockPath = path.join(baseDirectory, 'package-lock.json');
    if (fs.existsSync(yarnLockPath)) {
        return 'yarn';
    }
    else if (fs.existsSync(packageLockPath)) {
        return 'npm';
    }
    // Default to npm
    return 'npm';
};
export const askForCredentials = (bundle) => __awaiter(void 0, void 0, void 0, function* () {
    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-enter-credentials', {
        bundle: bundle,
    }));
    if (bundle.startsWith('https://nexus.magnolia-cms.com') ||
        bundle.startsWith('https://nexus.magnolia-cms.cn')) {
        logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('info-use-nexus-token'));
    }
    return inquirer.prompt([
        {
            type: 'input',
            name: 'username',
            message: 'Username',
        },
        {
            type: 'password',
            name: 'password',
            message: 'Password',
        },
    ]);
});
export function getWebAppConfigDefaultFolderLocation(tomcatFolder, instance) {
    return path.join(tomcatFolder, 'webapps', instance, 'WEB-INF', 'config', 'default');
}
export function getMicroprofileConfigLocation(tomcatFolder, instance) {
    return path.join(getWebAppConfigDefaultFolderLocation(tomcatFolder, instance), 'microprofile-config.properties');
}
export function copyDownloadedFile(bundle, file) {
    if (!(bundle.postCommand || []).some((command) => command === PostCommands.Extract ||
        command === PostCommands.ExtractAndOverride ||
        command === PostCommands.ExtractAndUnfold)) {
        const { dest = './' } = bundle;
        const destinationWithFileName = path.join(dest, path.normalize(file).split(path.sep).pop() ||
            'file' + new Date().getTime());
        fs.copySync(file, destinationWithFileName, { overwrite: true });
        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-file-downloaded-to', {
            path: path.resolve(destinationWithFileName),
        }));
        fs.rmSync(path.dirname(file), { recursive: true, force: true });
        return true;
    }
    return false;
}
