// @magnolia/cli-jumpstart-plugin/dist/lib/pj-helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import path from 'path';
import ora from 'ora';
import fs from 'fs-extra';
import { execa } from 'execa';
import { determinePackageManager, installCLI } from './helper.js';
import { i18nInstance, logger } from '../jumpstart-plugin.js';
export const handlePackageJSON = () => __awaiter(void 0, void 0, void 0, function* () {
    const pjPath = path.join(process.cwd(), 'package.json');
    if (!fs.existsSync(pjPath)) {
        yield initPackageJSON();
    }
    yield modifyPackageJSON(pjPath);
});
export const modifyPackageJSON = (pjPath) => __awaiter(void 0, void 0, void 0, function* () {
    let pj = JSON.parse(fs.readFileSync(pjPath, 'utf8'));
    let isMgnlCliInstalled = (pj === null || pj === void 0 ? void 0 : pj.dependencies) && (pj === null || pj === void 0 ? void 0 : pj.dependencies['@magnolia/cli']);
    if (!isMgnlCliInstalled) {
        isMgnlCliInstalled = yield installCLI();
        if (isMgnlCliInstalled) {
            pj = JSON.parse(fs.readFileSync(pjPath, 'utf8'));
        }
    }
    const nodeModulesPath = path.join(process.cwd(), 'node_modules');
    if (!fs.existsSync(nodeModulesPath)) {
        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-no-node_modules'));
        const pm = determinePackageManager();
        const npmISpinner = ora().start(i18nInstance.t('ora-start-running-pm-install', {
            packageManager: pm,
        }));
        try {
            yield execa(pm, ['install'], {
                buffer: true,
                cwd: process.cwd(),
            });
            npmISpinner.succeed(i18nInstance.t('ora-succeed-pm-install', {
                packageManager: pm,
            }));
        }
        catch (e) {
            npmISpinner.fail(i18nInstance.t('ora-fail-pm-install', {
                packageManager: pm,
                errorMsg: e.message,
            }));
            return;
        }
    }
    if (isMgnlCliInstalled) {
        const mgnlScript = {
            mgnl: 'node node_modules/@magnolia/cli',
        };
        const scriptKey = Object.keys(mgnlScript)[0];
        if (!pj.scripts) {
            pj.scripts = {};
        }
        if (!pj.scripts[scriptKey] ||
            pj.scripts[scriptKey] !== mgnlScript[scriptKey]) {
            pj.scripts[scriptKey] = mgnlScript[scriptKey];
        }
    }
    let changedTypeMsg;
    if (!pj.type) {
        pj.type = 'module';
    }
    else if (pj.type !== 'module') {
        changedTypeMsg = 'warn-changed-type';
        pj.type = 'module';
    }
    try {
        fs.writeFileSync(pjPath, JSON.stringify(pj, null, '\t'), 'utf8');
        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-add-mgnl-to-pj'));
        if (changedTypeMsg) {
            logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t(changedTypeMsg));
        }
    }
    catch (e) {
        logger === null || logger === void 0 ? void 0 : logger.error(i18nInstance.t('error-pj-mod-fail', { errorMsg: e.message }));
        return;
    }
});
export const initPackageJSON = () => __awaiter(void 0, void 0, void 0, function* () {
    const pjSpinner = ora().start(i18nInstance.t('ora-start-creating-pj'));
    try {
        yield execa(determinePackageManager(), ['init', '-y'], {
            buffer: true,
            cwd: process.cwd(),
        });
        pjSpinner.succeed(i18nInstance.t('ora-succeed-pj-created'));
    }
    catch (e) {
        pjSpinner.fail(i18nInstance.t('ora-fail-creating-pj', { errorMsg: e.message }));
        return;
    }
});
