// @magnolia/cli/dist/Cli.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { CreateError } from '@magnolia/cli-helper/general-utils';
import { Command } from 'commander';
import { createRequire } from 'module';
import { pathToFileURL } from 'url';
import path from 'path';
import { winstonLogger } from './lib/logger.js';
import TabCompletion from './lib/tab-completion.js';
import AddPlugin from './lib/add-plugin.js';
import chalk from 'chalk';
import { analytics } from './lib/analytics.js';
import fs from 'fs-extra';
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
import { getMgnlConfigPath } from '@magnolia/cli-helper/config';
import WatchMode from './lib/watch-mode.js';
import Init from './lib/init.js';
export let logger;
export class Cli {
    constructor(pluginManager, defaultConfig) {
        this.handleSigint = (plugin) => {
            let sigintHandled = false;
            let ignorePeriod = false;
            const stopPluginAndExit = () => __awaiter(this, void 0, void 0, function* () {
                if (!sigintHandled) {
                    ignorePeriod = true;
                    sigintHandled = true;
                    setTimeout(() => {
                        ignorePeriod = false;
                    }, 1000);
                    try {
                        yield plugin.stop();
                        this.stop();
                    }
                    catch (err) {
                        logger === null || logger === void 0 ? void 0 : logger.error('Error during graceful stop:', err);
                        process.exit(1);
                    }
                }
                else if (!ignorePeriod) {
                    logger === null || logger === void 0 ? void 0 : logger.error('Second SIGINT received. Forcefully terminating...');
                    process.exit(1);
                }
            });
            // Use .on to persist the listener in case a plugin/child process also catch it.
            process.on('SIGINT', stopPluginAndExit);
            // Return the stopAndExit function for testing
            return stopPluginAndExit;
        };
        this.pluginManager = pluginManager;
        this.defaultConfig = defaultConfig;
        this.customConfigPath = getMgnlConfigPath();
        logger = winstonLogger(defaultConfig, pkg.name, pkg.version);
    }
    importCustomConfig(configPath) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const module = yield import(configPath);
                this.customConfig = module.default;
            }
            catch (_a) {
                try {
                    const module = yield import(pathToFileURL(path.join(configPath)).href);
                    this.customConfig = module.default;
                }
                catch (e) {
                    if ((process.env.npm_lifecycle_event ||
                        process.env.npm_package_name) &&
                        process.env.npm_lifecycle_event !== 'npx') {
                        if (fs.existsSync(configPath)) {
                            logger.warn(`The config file at ${configPath} is corrupted and will not be used. See details below:`);
                        }
                        logger.warn(e.message);
                    }
                }
            }
        });
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.importCustomConfig(this.customConfigPath);
            yield analytics.init(this.customConfigPath);
        });
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const config = this.customConfig || this.defaultConfig;
            const label = chalk.bgGreen.white.bold('[mgnl 5]');
            const program = new Command();
            program
                .name(' ')
                .description(pkg.description)
                .version(`Magnolia CLI: ${pkg.version} (node.js: ${process.version})`, '-V, --Version')
                .showHelpAfterError()
                .addHelpText('beforeAll', label);
            if (((_a = config === null || config === void 0 ? void 0 : config.analytics) === null || _a === void 0 ? void 0 : _a.enabled) === true) {
                program.addHelpText('afterAll', "\nWe collect anonymized analytics to improve your experience.\nIf you prefer to opt out, set 'analytics.enabled' to 'false' in mgnl.config.js file.");
            }
            program.addHelpText('afterAll', '\nHave thoughts or suggestions? Let us know at cli-feedback@magnolia-cms.com');
            this.pluginManager.registerPlugins(config);
            if (this.customConfig !== undefined &&
                fs.existsSync(path.join(process.cwd(), 'package.json'))) {
                this.pluginManager.register(new AddPlugin({
                    version: pkg.version,
                    loggerConfig: pkg.logger,
                }));
            }
            this.pluginManager.register(new TabCompletion({
                version: pkg.version,
                loggerConfig: pkg.logger,
            }));
            const allPlugins = this.pluginManager.getPlugins();
            this.pluginManager.register(new WatchMode({
                version: pkg.version,
                loggerConfig: pkg.logger,
                allPlugins: allPlugins,
            }));
            this.pluginManager.register(new Init({
                version: pkg.version,
                loggerConfig: pkg.logger,
                allPlugins: allPlugins,
            }));
            for (const plugin of allPlugins) {
                const pluginCommand = program
                    .command(plugin.name)
                    .description(plugin.description)
                    .name(plugin.name)
                    .action((options, command) => __awaiter(this, void 0, void 0, function* () {
                    const logger = winstonLogger(config, plugin.constructor.name, plugin.version);
                    this.handleSigint(plugin);
                    try {
                        if (plugin.init) {
                            plugin.init(logger);
                        }
                        if (plugin.name !== 'add-plugin') {
                            yield analytics.send(pkg.version, config, plugin);
                        }
                        // eslint-disable-next-line no-console
                        console.log(label);
                        yield plugin.start(options, command.args);
                        if (plugin.name === 'jumpstart') {
                            yield analytics.init(this.customConfigPath);
                        }
                    }
                    catch (e) {
                        if (e instanceof CreateError) {
                            logger === null || logger === void 0 ? void 0 : logger.error(e.message);
                        }
                        else {
                            logger === null || logger === void 0 ? void 0 : logger.error(`Unexpected error occurred:\n${e.message}`);
                        }
                    }
                }));
                if (plugin.options) {
                    for (const option of plugin.options) {
                        pluginCommand.addOption(option);
                    }
                }
                if (plugin.usage) {
                    pluginCommand.usage(plugin.usage);
                }
                pluginCommand.version(`${plugin.constructor.name}: ${plugin.version}`, '-v, --version');
            }
            if (allPlugins.length === 0) {
                program.on('command:*', (e) => {
                    // eslint-disable-next-line no-console
                    console.log(`error: unknown command '${e.join(' ')}'\n`);
                    program.help();
                });
            }
            program.parse(process.argv);
        });
    }
    stop() {
        process.exit();
    }
}
