// @magnolia/cli-create-light-module-plugin/dist/create-light-module-plugin.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import path from 'path';
import { fileURLToPath } from 'url';
import { Option } from 'commander';
import { createRequire } from 'module';
import fs from 'fs-extra';
import inquirer from 'inquirer';
import { PluginTemplate } from '@magnolia/cli-plugin-template';
import { filterDirectories, filterMissingDirs } from './lib/helper.js';
import { resolveLightModulesPath, getMgnlConfig, initI18n, modifyConfig, findPrototypesPath, checkFlagsValue, getAvailablePrototypes, validateAndResolvePrototype, validateName, } from '@magnolia/cli-helper';
const requireFn = createRequire(import.meta.url);
const pkg = requireFn('./package.json');
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const cliPrototypesPath = path.resolve(__dirname, './resources/config/mgnl-cli-prototypes/light-module');
export let logger;
export let i18nInstance = {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    t(key, options) {
        return key;
    },
};
export default class CreateLightModulePlugin extends PluginTemplate {
    constructor(args) {
        super();
        this.pluginName = this.constructor.name;
        this.name = 'create-light-module';
        this.version = pkg.version;
        this.description = '';
        this.usage = '<name> [options]' +
            '\nParameter: <name> The name of the new light module. Avoid spaces and special characters since this name is used as folder name.';
        this.configUpdates = {
            sharedProps: {},
            pluginProps: {},
        };
        i18nInstance = initI18n(this.name, 'translation', path.join(__dirname, 'lib/locales'));
        this.description = i18nInstance.t('description');
        const prototypesPath = findPrototypesPath(args === null || args === void 0 ? void 0 : args.prototypeDir, cliPrototypesPath);
        const prototypes = getAvailablePrototypes(prototypesPath);
        const cmdOptionPrototype = prototypes
            ? 'option-prototype-description.withPrototypes'
            : 'option-prototype-description.noPrototypes';
        this.options = [
            new Option('-d, --directories [directories...]', i18nInstance.t('option-directories-description')).conflicts(['prototype', 'prototypeDir']),
            new Option('-md, --module-descriptor <version>', i18nInstance.t('option-moduleDescriptor-description')),
            new Option('-lmp, --light-modules-path [path]', i18nInstance.t('option-lightModulesPath-description')),
            new Option('-pd --prototype-dir <path>', i18nInstance.t('option-prototypeDir-description')).conflicts('directories'),
            new Option('-p, --prototype [name]', i18nInstance.t(cmdOptionPrototype, {
                prototypes,
                path: prototypesPath,
            })).conflicts('directories'),
        ];
        this.args = args || {};
    }
    init(winstonLogger) {
        return __awaiter(this, void 0, void 0, function* () {
            logger = winstonLogger;
        });
    }
    start(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = yield this.validateAndResolveArgs(options);
            yield this.createModule(params, options);
            yield modifyConfig(this.configUpdates, this.pluginName, logger);
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    validateAndResolveArgs(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const moduleName = validateName('light-module');
            checkFlagsValue(this.options, options);
            const config = yield getMgnlConfig(logger);
            const { lightModulesPath, newLightModulesPath } = yield resolveLightModulesPath(options, config, logger, true);
            if (newLightModulesPath) {
                this.configUpdates.sharedProps.lightModulesPath =
                    newLightModulesPath;
            }
            if (!config.lightModule) {
                this.configUpdates.sharedProps.lightModule = moduleName;
            }
            if (options.directories) {
                return {
                    lightModulesPath,
                    moduleName,
                    prototypeDir: cliPrototypesPath,
                    prototype: 'comprehensive',
                    directories: options.directories,
                };
            }
            const { prototype, prototypeDir, pluginUpdates } = yield validateAndResolvePrototype(options, this.args, cliPrototypesPath, logger);
            this.configUpdates.pluginProps = Object.assign(Object.assign({}, this.configUpdates.pluginProps), pluginUpdates);
            return {
                lightModulesPath,
                moduleName,
                prototypeDir,
                prototype,
                directories: options.directories,
            };
        });
    }
    createModule(vars, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { moduleName, lightModulesPath, prototype, prototypeDir, directories, } = vars;
            const srcPath = path.join(prototypeDir, prototype);
            const destPath = path.join(lightModulesPath, moduleName);
            const moduleExists = fs.existsSync(destPath);
            if (moduleExists) {
                logger === null || logger === void 0 ? void 0 : logger.warn(i18nInstance.t('warn-module-exists', {
                    name: moduleName,
                    path: destPath,
                }));
            }
            if (directories) {
                const { selectedDirs } = yield filterDirectories(srcPath, destPath, directories);
                for (const dir of selectedDirs) {
                    fs.copySync(path.join(srcPath, dir), path.join(destPath, dir), {
                        filter: (src) => path.basename(src) !== '.keep',
                    });
                }
                fs.copySync(path.join(srcPath, 'README.md'), path.join(destPath, 'README.md'));
                if (!moduleExists) {
                    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-module-created', {
                        name: moduleName,
                        path: destPath,
                    }));
                }
                if (selectedDirs.length > 0) {
                    logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-missing-modules-created', {
                        modules: selectedDirs.join(', '),
                        path: destPath,
                    }));
                }
            }
            else {
                const missingDirs = moduleExists
                    ? filterMissingDirs(srcPath, destPath)
                    : [];
                let confirm = true;
                if (missingDirs.length > 0) {
                    const answer = yield inquirer.prompt([
                        {
                            type: 'confirm',
                            name: 'confirm',
                            message: i18nInstance.t('inquirer-prompt-add-missing-dir', {
                                moduleName: moduleName,
                                missingDirs: missingDirs.join(', '),
                                prototype,
                            }),
                        },
                    ]);
                    confirm = answer.confirm;
                }
                if (confirm) {
                    fs.copySync(srcPath, destPath, {
                        filter: (src) => path.basename(src) !== '.keep',
                    });
                    if (!moduleExists) {
                        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-module-created', {
                            name: moduleName,
                            path: destPath,
                        }));
                    }
                    if (missingDirs.length > 0) {
                        logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-missing-modules-created', {
                            modules: missingDirs.join(', '),
                            path: destPath,
                        }));
                    }
                }
            }
            const i18nProperties = path.join(destPath, 'i18n', '__name__-messages_en.properties');
            if (fs.existsSync(i18nProperties)) {
                fs.renameSync(i18nProperties, path.join(destPath, 'i18n', `${moduleName}-messages_en.properties`));
            }
            this.createModuleDescriptor(destPath, moduleExists, options.moduleDescriptor);
        });
    }
    createModuleDescriptor(destPath, moduleExists, moduleDescriptor) {
        if (!moduleExists) {
            const content = `version: ${moduleDescriptor || '1.0.0'}\n`;
            fs.writeFileSync(path.join(destPath, 'module.yaml'), content, 'utf8');
        }
        else if (moduleDescriptor) {
            const moduleDescriptorFile = path.join(destPath, 'module.yaml');
            if (fs.existsSync(moduleDescriptorFile)) {
                logger.warn(i18nInstance.t('warn-module-descriptor-exists'));
            }
            else {
                const content = `version: ${moduleDescriptor}\n`;
                fs.writeFileSync(moduleDescriptorFile, content, 'utf8');
                logger === null || logger === void 0 ? void 0 : logger.info(i18nInstance.t('info-module-descriptor-created', {
                    path: destPath,
                }));
            }
        }
    }
}
